src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs

```diff
@@ -287,4 +289,152 @@ public async Task SendAsync_WithZeroLengthHeaderName_Throws()
+
+            LoopbackServerFactory.CreateClientAndServerAsync(
+                async uri =>
+                {
+                    using var requestMessage = new HttpRequestMessage(HttpMethod.Get, uri) { Version = VersionFromUseHttp2 };
+
+                    foreach ((string name, string[] values) in headers)
+                    {
+                        requestMessage.Headers.Add(name, values);
+                    }
+
+                    using HttpClient client = CreateHttpClient();
+
+                    if (expectSuccess)
+                    {
+                        await client.SendAsync(requestMessage);
+                    }
+                    else
+                    {
+                        await Assert.ThrowsAsync<HttpRequestException>(() => client.SendAsync(requestMessage));
+                    }
+                },
+                async server =>
+                {
+                    HttpRequestData requestData = null;
+                    try
+                    {
+                        requestData = await server.HandleRequestAsync();
+                    }
+                    catch (Exception)
+                    {
+                        if (expectSuccess)
+                            throw;
+                    }
+
+                    if (!expectSuccess)
+                        return;
+
+                    Assert.All(requestData.Headers,
+                        h => Assert.False(h.HuffmanEncoded, "Expose raw decoded bytes once HuffmanEncoding is supported"));
+
+                    Encoding valueEncoding = Encoding.GetEncoding("ISO-8859-1");
+
+                    foreach ((string name, string[] values) in headers)
+                    {
+                        byte[] valueBytes = valueEncoding.GetBytes(string.Join(", ", values));
+
+                        Assert.Single(requestData.Headers,
+                            h => h.Name.Equals(name, StringComparison.OrdinalIgnoreCase) && h.Raw.AsSpan().IndexOf(valueBytes) != -1);
+                    }
+                })
+                .GetAwaiter().GetResult();
+
+            return RemoteExecutor.SuccessExitCode;
+        }
+
+        protected int SendAsync_ReceiveNonAsciiCharacters_Inner(string useAppCtxSwitchInner, string switchValueInner, string unicodeInner)
+        {
+            bool allowLatin1 = AllowLatin1Headers(useAppCtxSwitchInner, switchValueInner);
+            bool unicode = bool.Parse(unicodeInner);
+            bool expectSuccess = allowLatin1 && !unicode;
+
+            LoopbackServerFactory.CreateClientAndServerAsync(
+                async uri =>
+                {
+                    using var requestMessage = new HttpRequestMessage(HttpMethod.Get, uri) { Version = VersionFromUseHttp2 };
+                    using HttpClient client = CreateHttpClient();
+
+                    using HttpResponseMessage response = await client.SendAsync(requestMessage);
+
+                    if (!expectSuccess)
```

Comment:
It doesn't throw in this case?

Reply:
Not in the receive test. It's actually only a regression test (will also succeed without my product code changes).

In this test, `expectSuccess` is only an indicator whether we should validate for Latin-1. Will think if I can make it easier to understand.